### DISCUSSION TITLE
[v4.1] How to extends QueryBuilder

### DISCUSSION THREAD

**Discussion by maidiries** [adonisjs/core#2484] [Help]
I need add a synconously mothod to QueryBuilder

User Model:
```JavaScript
class User extends Model {
  posts(){
    return this.hasMany("App/Models/Posts")
  }
}
```

addFilters method:
```JavaScript
addFilters(params){
  if( params.subject ){
    this.where("subject", params.subject)
  }
  if( params.date_from ){
    this.where("created_at", ">=", params.date_from)
  }
  ...
}
```

Finally, I will run in Controller:
```JavaScript
const posts = await User.posts().addFilters( request.get() ).paginate(2, 2)
```

What is the best way to implement "addFilters" functionality?

**Comment by thetutlage** [comment]
Use query scopes. Docs here https://adonisjs.com/docs/4.0/lucid#_query_scopes

**Reply by maidiries** [reply]
But I need it for all models. In the example before, I have another model like Locations that I would like to add the same addFilters method

**Reply by thetutlage** [reply]
You can use class inheritance for that

**Reply by maidiries** [reply]
Where is the QueryBuilder class to extend it?

I found this solution:
```JavaScript
const Knex = use('knex')

Knex.QueryBuilder.extend('addFilters', function(params){
  if( params.subject ){
    this.where("subject", params.subject)
  }
  if( params.date_from ){
    this.where("created_at", ">=", params.date_from)
  }
  ...
})
```
It is Adonis friendly?

**Comment by artem-a** [comment]
Hi @maidiries you can combine 
[Scopes](https://adonisjs.com/docs/4.1/lucid#_query_scopes) & [Traits](https://adonisjs.com/docs/4.1/traits)

Describe your filters using scopes inside the traits and then connect to the specific models

**Reply by artem-a** [reply]
Did you read [this](https://adonisjs.com/docs/4.1/traits#_extending_query_builder)?
